src/2_branch.js

```diff
@@ -6,322 +6,740 @@ goog.provide('Branch');
+ *           identity:       'BranchUser' // Unique string that identifies the user
+ *      }
+ * );
+ * ```
+ *
+ * **Note:** `Branch.init` must be called prior to calling any other Branch functions.
+ * ___
  */
 Branch.prototype['init'] = function(app_id, callback) {
-	if (this.initialized) { return callback(utils.message(utils.messages.existingInit)); }
-	this.initialized = true;
+	callback = callback|| function() { };
+	if (this.initialized) {
+		return callback(utils.message(utils.messages.existingInit));
+	}
 
-	callback = callback || function() {};
 	this.app_id = app_id;
-
 	var self = this, sessionData = utils.readStore();
-	if (sessionData && !sessionData['session_id']) { sessionData = null; }
 
-	if (sessionData) {
-		this.session_id = sessionData['session_id'];
-		this.identity_id = sessionData['identity_id'];
-	}
+	var setBranchValues = function(data) {
+		self.session_id = data['session_id'];
+		self.identity_id = data['identity_id'];
+		self.sessionLink = data['link'];
+		self.initialized = true;
+	};
 
-	if (sessionData && !utils.hashValue('r')) {
-		callback(null, sessionData);
+	if (sessionData  && sessionData['session_id']) {
+		setBranchValues(sessionData);
+		callback(null, utils.whiteListSessionData(sessionData));
 	}
 	else {
-		this.api(resources._r, {}, function(err, browser_fingerprint_id) {
-			if (err) { callback(err); }
-			else {
-				self.api(resources.open, {
-					"link_identifier": utils.hashValue('r'),
-					"is_referrable": 1,
-					"browser_fingerprint_id": browser_fingerprint_id
-				}, function(err, data) {
-					if (err) { callback(err); }
-					else {
-						self.session_id = data['session_id'];
-						self.identity_id = data['identity_id'];
-						utils.store(data);
-						callback(null, data);
-					}
-				});
-			}
-		})
+		this._api(resources._r, {}, function(err, browser_fingerprint_id) {
+			self._api(resources.open, {
+				"link_identifier": utils.hashValue('r'),
+				"is_referrable": 1,
+				"browser_fingerprint_id": browser_fingerprint_id
+			}, function(err, data) {
+				setBranchValues(data);
+				utils.store(data);
+				callback(err, utils.whiteListSessionData(data));
+			});
+		});
 	}
 };
 
 /**
- * @param {function|null} callback
+ * @function Branch.data
+ * @param {function|null} callback - _optional_ - callback to read the session data.
+ *
+ * Returns the same session information and any referring data, as
+ * `Branch.init`, but does not require the `app_id`. This is meant to be called
+ * after `Branch.init` has been called if you need the session information at a
+ * later point.
+ * If the Branch session has already been initialized, the callback will return
+ * immediately, otherwise, it will return once Branch has been initialized.
+ * ___
+ */
+Branch.prototype['data'] = function(callback) {
+	callback = callback || function() { };
+	callback(null, utils.whiteListSessionData(utils.readStore()));
```

Comment:
let's put this inside the queue, so that if you do branch.init() branch.data() it calls data *after* init finishes